### PR TITLE
Initial xclbinutil PSKernel support for xclbin containers

### PR DESCRIPTION
**Pull request by rozumx**
Initial support to datamine a shared PSKernel library (e.g., ELF format) to discover the PS Kernel, its arguments and types, and supporting functions (e.g., _init and _fini).

Work Done
---------
+ Added new MEM_TOPOLOGY type MEM_PS_KERNEL
+ Updated xclbinutil/CMakeLists.txt with new build and test sources
+ Opportunistically beautified modified sources to match google style.
+ Created ElfUtilities to extra function and argument metdata from a shared library
+ Update KernelUtilties to:
a. Validate there is only 1 kernel and at max 1 _init or _fini helper functions
b. Validate the that _init and _fini are well formed.
c. Validate that the _init and _fini have the same base name as the kernel
d. Transpose the functions to the JSON ps-kernel schema.
e. Create a MEM_PS_KERNEL MEM_TOPOLOGY entry if one doesn't exist
f. Refactored code to be both FIXED and PS Kernel friendly
+ Updated the Soft and Fixed Kernel regress tests regarding alternate IP_LAYOUT output.
+ Updated Fixed Kernel regresion test regarding using --add-pskernel option
+ Added regression tests for PS kernel.  These tests will examine the shared library ELF format and populate the xclbin containing accordingly

### CODE REVIEW

File: src/runtime_src/tools/xclbinutil/XclBinClass.cxx [hunk 1]
```diff
@@ -16,96 +16,103 @@
...
-#include <boost/format.hpp>
+#include <stdexcept>
+#include <stdlib.h>
 
-#include "XclBinUtilities.h"
-#include "KernelUtilities.h"
-namespace XUtil = XclBinUtilities;
+// Constant data
+static const std::string MIRROR_DATA_START("XCLBIN_MIRROR_DATA_START");
+static const std::string MIRROR_DATA_END("XCLBIN_MIRROR_DATA_END");
```
Comment (stsoe) [inline src/runtime_src/tools/xclbinutil/XclBinClass.cxx:41]: lower case variable names is better, in general ALL_UPPER_CASE should be reserved for macros, which we shouldn't really use ;_)

File: src/runtime_src/tools/xclbinutil/XclBinClass.cxx [hunk 2]
```diff
@@ -1585,62 +1605,62 @@ XclBin::setKeyValue(const std::string & _keyValue)
...
-         bKeyFound = true;
-         XUtil::QUIET(std::string("Removing key '") + _sKey + "'");
-         keyValues.erase(keyValues.begin() + index);
-         break;
-      }
+  pSection->getPayload(ptKeyValueMetadata);
+
+  XUtil::TRACE_PrintTree("KEYVALUE:", ptKeyValueMetadata);
+  boost::property_tree::ptree ptKeyValues = ptKeyValueMetadata.get_child("keyvalue_metadata");
+  std::vector<boost::property_tree::ptree> keyValues = XUtil::as_vector<boost::property_tree::ptree>(ptKeyValues, "key_values");
```
Comment (stsoe) [inline src/runtime_src/tools/xclbinutil/XclBinClass.cxx:1625]: why repeat the type?  auto?

File: src/runtime_src/tools/xclbinutil/XclBinClass.cxx [hunk 3]
```diff
@@ -1585,62 +1605,62 @@ XclBin::setKeyValue(const std::string & _keyValue)
...
+  }
 
-   // Now create a new tree to add back into the section
-   boost::property_tree::ptree ptKeyValuesNew;
-   for (auto keyvalue : keyValues) {
-     ptKeyValuesNew.push_back(std::make_pair("", keyvalue));
-   }
+  // Now create a new tree to add back into the section
+  boost::property_tree::ptree ptKeyValuesNew;
+  for (auto keyvalue : keyValues) {
```
Comment (stsoe) [inline src/runtime_src/tools/xclbinutil/XclBinClass.cxx:1645]: auto& ?

File: src/runtime_src/tools/xclbinutil/XclBinClass.cxx [hunk 4]
```diff
@@ -1649,28 +1669,32 @@ XclBin::reportInfo(std::ostream &_ostream, const std::string & _sInputFile, bool
...
+  const std::string delimiters = ":";  
 
   // Working variables
   std::string::size_type pos = 0;
   std::string::size_type lastPos = 0;
   std::vector<std::string> tokens;
 
-  // Parse the string until the entire string has been parsed or 3 tokens have been found
+  // Parse the string until the entire string has been parsed or MAX_TOKENS tokens have been found
+  constexpr size_t MAX_TOKENS = 3;
```
Comment (stsoe) [inline src/runtime_src/tools/xclbinutil/XclBinClass.cxx:1688]: I am vary of UPPER_CASE variables subject to preprocessor

**Review by stsoe — CHANGES_REQUESTED** [review]
Some inline minor comments, I didn't review in great detail.

What is src/runtime_src/tools/xclbinutil/unittests/PSKernel/pskernel.hex  ?  Some hex dump of binary data?

**Comment by rozumx** [comment]
> What is src/runtime_src/tools/xclbinutil/unittests/PSKernel/pskernel.hex ? Some hex dump of binary data?
To temporary address the binary image check-in limitation, the pskernel.so image was converted to a hex dump.

I am currently working on updating the XRT build environment to enable ARM cross compilation for built in PS kernels.  When I complete this work, I will be updating this unit test to build the pskernel.so directly from the source (pskernel.cpp).   

Note: If you look at the pskernel.cpp file, you will see the commands needed to compile this simple wrapper file into a pskernel.so shared library.

File: src/runtime_src/tools/xclbinutil/ElfUtilities.cxx [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+    while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr) {
+        result += buffer.data();
+    }
+    return result;
+}
+
+
+
+void
```
[the remark this answers is not in the document]
Comment (rozumx) [reply src/runtime_src/tools/xclbinutil/ElfUtilities.cxx:51]: I am more than happy to match this change.   
For some reason I thought multiple copies (possibly 2) of a collection was made when it was a return value.   To test this theory, I created a simple class that counted the number of times the constructor was called.
class Counter
```
{
 public:
  Counter() {count++;}
    int currentValue() { return count;}
 private:
  static int count;
};
```
And returned this value from the function call:
```
  Counter counter;
  return counter;
```
And printed out the results:
```
  Counter counter = dataMindExportedFunctionsObjdump(elfLibrary, kernelSignatures);
  std::cout << "Counter: "  << counter.currentValue() << std::endl;
```
To my pleasant surprise, the number reported is 1.
`Counter: 1`

Thanks again for catching this.

File: src/runtime_src/tools/xclbinutil/ElfUtilities.cxx [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+  const std::string output = exec(cmd);
+
+  std::vector<std::string> entries;
+  boost::split(entries, output, boost::is_any_of("\n"), boost::token_compress_on);
+
+  // Look for the following function attributes in the '.text' section
+  // g - Global
+  // F - Function
+
+  for (auto entry : entries) {
```
[the remark this answers is not in the document]
Comment (rozumx) [reply src/runtime_src/tools/xclbinutil/ElfUtilities.cxx:78]: done

File: src/runtime_src/tools/xclbinutil/ElfUtilities.cxx [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    // Look for the ' .text' entry.
+    std::size_t textIndex = entry.find(" .text");
```
[the remark this answers is not in the document]
Comment (rozumx) [reply src/runtime_src/tools/xclbinutil/ElfUtilities.cxx:80]: Fixed this issue and others like this in the code base.

File: src/runtime_src/tools/xclbinutil/ElfUtilities.cxx [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    if (textIndex == std::string::npos)
+      continue;
+
+    // Examine the flags looking for a global function.
+    std::size_t flagIndex = entry.find(" ");
```
[the remark this answers is not in the document]
Comment (rozumx) [reply src/runtime_src/tools/xclbinutil/ElfUtilities.cxx:85]: done

File: src/runtime_src/tools/xclbinutil/unittests/PSKernel/ip_layout_expected.json [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+                "m_interrupt_id": "0",
+                "m_ip_control": "AP_CTRL_CHAIN",
+                "m_base_address": "0x80000",
+                "m_name": "vadd:vadd_1"
+            },
+            {
+                "m_type": "IP_PS_KERNEL",
+                "properties": "0x0",
+                "m_base_address": "not_used",
+                "m_name": "kernel0:0"
```
[the remark this answers is not in the document]
Comment (rozumx) [reply src/runtime_src/tools/xclbinutil/unittests/PSKernel/ip_layout_expected.json:17]: All ps kernel instances will now be prefixed with the "scu_" value.

File: src/runtime_src/tools/xclbinutil/XclBinClass.cxx (hunk 2 above)
Comment (rozumx) [reply src/runtime_src/tools/xclbinutil/XclBinClass.cxx:1625]: I understand fully what auto does and I do use it frequency when I am working with iterators.   

That said, I do see the redundancy here and have updated the return type to be auto (as well as other places in the code base).

File: src/runtime_src/tools/xclbinutil/XclBinClass.cxx (hunk 3 above)
Comment (rozumx) [reply src/runtime_src/tools/xclbinutil/XclBinClass.cxx:1645]: Fixed here and in other locations of this file.

File: src/runtime_src/tools/xclbinutil/XclBinClass.cxx (hunk 4 above)
Comment (rozumx) [reply src/runtime_src/tools/xclbinutil/XclBinClass.cxx:1688]: Refactored to use camel case.

File: src/runtime_src/tools/xclbinutil/ElfUtilities.cxx [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    std::string functionSig = entry.substr(baseIndex + sizeof("Base"));
+    boost::algorithm::trim(functionSig);
+
+    kernelSignatures.push_back(functionSig);
+  }
+}
+
+
+
+void
```
[the remark this answers is not in the document]
Comment (rozumx) [reply src/runtime_src/tools/xclbinutil/ElfUtilities.cxx:110]: Fixed

File: src/runtime_src/tools/xclbinutil/unittests/PSKernel/embedded_metadata.xml [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+              <rtlPort name="v2_buffer_q0" object="v2_buffer" protocol="ap_memory"/>
+              <rtlPort name="vout_buffer_d0" object="vout_buffer" protocol="ap_memory"/>
+            </module>
+            <module name="vadd_Pipeline_write" instName="grp_vadd_Pipeline_write_fu_193" type="NonDataflowHS">
+              <rtlPort name="m_axi_gmem_AWVALID" object="gmem" protocol="m_axi"/>
+              <rtlPort name="sext_ln92" object="sext_ln92" protocol="ap_none"/>
+              <rtlPort name="trunc_ln67_1" object="trunc_ln67_1" protocol="ap_none"/>
+              <rtlPort name="vout_buffer_q0" object="vout_buffer" protocol="ap_memory"/>
+            </module>
+          </module>
```
[the remark this answers is not in the document]
Comment (rozumx) [reply src/runtime_src/tools/xclbinutil/unittests/PSKernel/embedded_metadata.xml:52]: For this use-case, I simple copied the contents of a previous flow (e.g., Fix PS Kernel).  This data is NOT used by PS Kernel, but it is existing contents that should NOT be lost when adding PS kernels, hence they are present.

File: src/runtime_src/tools/xclbinutil/unittests/PSKernel/embedded_metadata_expected.xml [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+              <rtlPort name="v2_buffer_q0" object="v2_buffer" protocol="ap_memory"/>
+              <rtlPort name="vout_buffer_d0" object="vout_buffer" protocol="ap_memory"/>
+            </module>
+            <module name="vadd_Pipeline_write" instName="grp_vadd_Pipeline_write_fu_193" type="NonDataflowHS">
+              <rtlPort name="m_axi_gmem_AWVALID" object="gmem" protocol="m_axi"/>
+              <rtlPort name="sext_ln92" object="sext_ln92" protocol="ap_none"/>
+              <rtlPort name="trunc_ln67_1" object="trunc_ln67_1" protocol="ap_none"/>
+              <rtlPort name="vout_buffer_q0" object="vout_buffer" protocol="ap_memory"/>
+            </module>
+          </module>
```
[the remark this answers is not in the document]
Comment (rozumx) [reply src/runtime_src/tools/xclbinutil/unittests/PSKernel/embedded_metadata_expected.xml:52]: Great to know.    I will validate that the:
1. XRT code base is not using this data.
2. HW Emulation is not using this data.

If that turns out to be the case, then I will instruct the Vitis flow to remove this contents.

File: src/runtime_src/tools/xclbinutil/XclBinClass.cxx (hunk 1 above)
Comment (rozumx) [reply src/runtime_src/tools/xclbinutil/XclBinClass.cxx:41]: Renamed the string to be camelCase.

**Comment by rozumx** [comment]
retest this please

**Comment by rozumx** [comment]
> Shouldn't this be a private target link library for NOT WIN32 ?

Yes it should and it was my oversight on my part.

**Comment by rozumx** [comment]
retest this please

**Review by AShivangi — APPROVED** [review]

**Comment by rozumx** [comment]
retest this please

**Review by stsoe — APPROVED** [review]

**Comment by dayeh-xilinx** [comment]
retest this please.

**Pull request merged**